### PR TITLE
Run database updates after configuration import

### PR DESCRIPTION
**Pull request by nathandentzau**
Running database updates before configuration import can cause two potential issues:

1) Configuration changes made in hook_update_N() or post update hooks can be overwritten by configuration import.
2) Running hook_update_N() or post update hooks that rely on new configuration will fail.

There are no foreseen consequences changing the order of operations here.

### CODE REVIEW

**Comment by adamzimmermann** [comment]
> Configuration changes made in hook_update_N() or post update hooks can be overwritten by configuration import.

This would be an issue during the subsequent deployment anyways, so this change would just delay the problem. Unless of course we were adding said configuration to the ignored list in this process.

> Running hook_update_N() or post update hooks that rely on new configuration will fail.

The same could be said for importing configuration that relies upon new database table structures that were altered via the update hook.

I've grappled with this before and always forget the reasoning for both sides so I took some notes last time I worked on something like this.

Here is what I had written down.
```
updb before config-import
This ensures new/updated database structures are in place to store configuration.
vs.
config-import before updb
This ensure fields are created before update hooks that use them are ran.
```

So all that to say, I'm not sure there is a right answer. I guess it depends on what you are doing in your update hooks?

I would love to be proven wrong though, and for us to come to a definitive answer that works for all cases.

**Comment by nathandentzau** [comment]
> This would be an issue during the subsequent deployment anyways, so this change would just delay the problem. Unless of course we were adding said configuration to the ignored list in this process.

This is true. 

> The same could be said for importing configuration that relies upon new database table structures that were altered via the update hook.

I don't think that is correct. Drupal config import wouldn't fail if a custom table is not present. It doesn't fail when config references an entity id when it does not exist in the database (e.g. exporting config for a custom block entity that does not exist in the production database). [Check out this article on configuration entity dependencies][]. Config does not set dependencies on content. Importing configuration does not alter content or need to know if content exists in a system (e.g. being able to install a new instance of Drupal from existing configuration). 

[Check out this article on configuration entity dependencies]: https://www.drupal.org/docs/8/api/configuration-api/configuration-entity-dependencies

> This ensures new/updated database structures are in place to store configuration.

This note is not correct IMO. Configuration is not stored in custom schema. Configuration is _only_ stored in the config table as a serialized value. The YAML files are stored in the database as a serialized PHP array. Any "config" stored in custom tables is not considered config in Drupal as the configuration management system has no context of custom schema.

I opened this PR because my Tugboat build is broken due to the order of operations in this role being incorrect. I've added a new field to a taxonomy vocabulary and then I have an update hook that's automatically setting a value for existing content. In this role's current state, I would have to create two different pull requests to deploy this feature because the configuration would need to exist before my database updates are ran. However, if the config import happened before the database updates then I could deploy this feature in a single PR. I think this example in itself is enough to show that the order in which we do database updates and configuration importing is not correct. 

Here's the Tugboat build: https://dashboard2.tugboat.qa/5c87e883efdbcd0001c8e47c

**Comment by nathandentzau** [comment]
Well the comment submitted before I could finish. What I wanted to end with is that the configuration import and database schema updates are not perfect, and it will most likely not work 100% of the time.

However in the order of:

1. Cache clear
2. Config import
3. Database updates
4. Entity updates

It will get you there about 99% of the time.

**Comment by adamzimmermann** [comment]
> This note is not correct IMO. Configuration is not stored in custom schema. Configuration is only stored in the config table as a serialized value. The YAML files are stored in the database as a serialized PHP array. Any "config" stored in custom tables is not considered config in Drupal as the configuration management system has no context of custom schema.

Ya, I may be conflating some D7 constructs with D8 things here, as there are important differences that you point out. I wish I had taken more detailed notes of what I ran into that had me put it this way once upon a time.

> I opened this PR because my Tugboat build is broken due to the order of operations in this role being incorrect. I've added a new field to a taxonomy vocabulary and then I have an update hook that's automatically setting a value for existing content. In this role's current state, I would have to create two different pull requests to deploy this feature because the configuration would need to exist before my database updates are ran. However, if the config import happened before the database updates then I could deploy this feature in a single PR. I think this example in itself is enough to show that the order in which we do database updates and configuration importing is not correct.

That's hard to refute haha.

tl;dr You have me convinced! Hopefully we don't run into any issues with this and this solves it!

**Review by adamzimmermann — APPROVED** [review]
Now I need to go update the RPX deploy script too!

**Comment by markdorison** [comment]
> Now I need to go update the RPX deploy script too!

@adamzimmermann If you all used this ansible role to deploy you wouldn't need to!

**Comment by nathandentzau** [comment]
:shotsfired:

**Comment by nathandentzau** [comment]
@markdorison are you good with this being merged in or do you want others in weigh in on this?

**Review by markdorison — APPROVED** [review]
All of the above discussion makes sense to me. I believe all of our repos _except for CHQ_ pin the role to a specific version, so this change won't roll out automatically except for there. For that reason, it is relatively safe to merge this change.
